src/htcondorce/web.py

```diff
@@ -19,21 +21,56 @@
 _initialized = None
 _loader = None
 _view = None
+_plugins = []
 g_is_multice = False
 OK_STATUS = '200 OK'
 
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+
+def validate_plugin(name, plugin):
+    try:
+        for regex, callback in plugin.urls:
+            if not hasattr(regex, "match"):
+                log.warn("plugin %s: %r is not a compiled regex (real type: %s)" % (name, regex, type(regex)))
+                return False
+            if not callable(callback):
+                log.warn("plugin %s: regex %r does not have a valid callback (real type: %s)"
+                      % (name, regex.pattern, type(callback)))
+                return False
+    except (ValueError, AttributeError) as err:
+        log.warn("plugin %s: %s" % (name, err))
+        return False
+    return True
+
+
 def check_initialized(environ):
     global _initialized
     global _loader
     global _cp
+    global _plugins
     global htcondor
+
     if not _initialized:
         if 'htcondorce.templates' in environ:
             _loader = genshi.template.TemplateLoader(environ['htcondorce.templates'], auto_reload=True)
         else:
             _loader = genshi.template.TemplateLoader('/usr/share/condor-ce/templates', auto_reload=True)
         ce_config = environ.get('htcondorce.config', '/etc/condor-ce/condor_config')
         htcondor = htcondorce.web_utils.check_htcondor()
+
+        plugins_dir = htcondor.param.get("HTCONDORCE_VIEW_PLUGINS_DIR", "/usr/share/condor-ce/ceview-plugins")
+        if os.path.isdir(plugins_dir):
+            for filename in sorted(os.listdir(plugins_dir)):
+                if not filename.endswith(".py"):
+                    continue
+                name = filename[:-3]
+                plugin = imp.load_source(name, os.path.join(plugins_dir, filename))
```

Comment:
The imp interface has been deprecated in favor of the importlib module. https://docs.python.org/3/library/importlib.html#module-importlib

In fact, I can't find the load_source function in the python 3's documentation of the old imp module.

Reply:
Yeah, we'll have to fix that up when we port this to Python 3.